component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/utils/TokenExchangeUtils.java

```diff
@@ -749,4 +856,71 @@ private static boolean validateUsingCertificate(SignedJWT signedJWT, IdentityPro
         // At this point 'verifier' will never be null;
         return signedJWT.verify(verifier);
     }
+
+    public static User getLocalUser(OAuthTokenReqMessageContext tokReqMsgCtx,
+                                    String subjectIdentifier) throws IdentityOAuth2Exception {
+
+        User localUser = null;
+        AbstractUserStoreManager userStoreManager = getUserStoreManager(tokReqMsgCtx);
+        try {
+             if(userStoreManager.isExistingUser(subjectIdentifier)) {
+                 localUser = userStoreManager.getUser(null, subjectIdentifier);
+             }
+        } catch (UserStoreException e) {
+            handleException("Error while resolving local user for subject: " + subjectIdentifier);
+        }
+        return localUser;
+    }
+
+    private static String resolveSubjectIdentifier(JWTClaimsSet claimsSet, String[] lookupAttributes)
+            throws IdentityOAuth2Exception {
+
+        String subjectIdentifier = null;
+        for (String lookupAttribute : lookupAttributes) {
+            Object subjectIdentifierObj = claimsSet.getClaim(lookupAttribute);
+            if (subjectIdentifierObj instanceof String) {
+                subjectIdentifier = (String) subjectIdentifierObj;
+            }
+
+            if (lookupAttribute.equals("email") && StringUtils.isBlank(subjectIdentifier)) {
+                String regex = "^(.+)@(.+)$";
+                Pattern pattern = Pattern.compile(regex);
+                Matcher matcher = pattern.matcher(claimsSet.getSubject());
+                if (matcher.matches()) {
+                    subjectIdentifier = claimsSet.getSubject();
+                }
+            }
```

Comment:
why do we only consider email here? also better iterate all the lookup params for specific claims and then go to match the subject. eg. if both email and mobile are configured as lookup, flow would be check for email attribute, check for mobile attribute, check if subject match email regex, check if subject match mobile regex (whichever succeed first will proceed)

Reply:
This needs to be fixed. however due to the urgency to merge the pr we will proceed for now since this is not a blocking issue. I will send a followup PR to fix this issue.